pkg/speaker/vip/keepalived.go

```diff
@@ -0,0 +1,192 @@
+package vip
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-logr/logr"
+	"github.com/openelb/openelb/pkg/constant"
+	"github.com/openelb/openelb/pkg/speaker"
+	appv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"strings"
+)
+
+type KeepAlive struct {
+	log       logr.Logger
+	clientset *clientset.Clientset
+	cm        *corev1.ConfigMap
+	conf      *KeepAliveConfig
+}
+
+type KeepAliveConfig struct {
+	Args  []string
+	Image string
+}
+
+func (k *KeepAlive) SetBalancer(configMap string, nexthops []corev1.Node) error {
+	ip := strings.SplitN(configMap, ":", 2)
+	k.cm = &corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ConfigMap",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      constant.OpenELBConfigMap,
+			Namespace: constant.OpenELBNamespace,
+		},
+		Data: map[string]string{
+			ip[0]: ip[1],
+		},
+	}
+	k.log.Info(fmt.Sprintf("cm %s", k.cm.String()))
+	var err error
+	if oldCm, err := k.clientset.CoreV1().ConfigMaps(k.cm.ObjectMeta.Namespace).Get(context.TODO(), k.cm.ObjectMeta.Name, metav1.GetOptions{}); errors.IsNotFound(err) {
+		k.cm, err = k.clientset.CoreV1().ConfigMaps(k.cm.ObjectMeta.Namespace).Create(context.TODO(), k.cm, metav1.CreateOptions{})
+		k.log.Info(fmt.Sprintf("create cm %s", k.cm.ObjectMeta.Name))
+		if err != nil {
+			k.log.Error(err, "create cm error")
+		}
+	} else {
+		for oldk, oldv := range oldCm.Data {
+			if _, ok := k.cm.Data[oldk]; !ok {
+				k.cm.Data[oldk] = oldv
+			}
+		}
+		k.cm, err = k.clientset.CoreV1().ConfigMaps(k.cm.Namespace).Update(context.TODO(), k.cm, metav1.UpdateOptions{})
+	}
+
+	return err
+}
+
+func (k *KeepAlive) DelBalancer(configMap string) error {
+	var err error
+	if _, err = k.clientset.CoreV1().ConfigMaps(k.cm.ObjectMeta.Namespace).Get(context.TODO(), k.cm.ObjectMeta.Name, metav1.GetOptions{}); err == nil {
+		ip := strings.SplitN(configMap, ":", 2)
+		delete(k.cm.Data, ip[0])
+		k.cm, err = k.clientset.CoreV1().ConfigMaps(k.cm.ObjectMeta.Namespace).Update(context.TODO(), k.cm, metav1.UpdateOptions{})
+	}
+	return err
+}
+
+func (k *KeepAlive) Start(stopCh <-chan struct{}) error {
+	daemonSetclient := k.clientset.AppsV1().DaemonSets(constant.OpenELBNamespace)
+	var privileged = true
```

Comment:
I think you prefer to use `daemonSetClient` instead of `daemonSetclient`.

Reply:
ok